graylog2-server/src/main/java/org/graylog2/alerts/AlertServiceImpl.java

```diff
@@ -17,122 +17,113 @@
+        this.alertConditionFactory = alertConditionFactory;
+        final String collectionName = AlertImpl.class.getAnnotation(CollectionName.class).value();
+        final DBCollection dbCollection = mongoConnection.getDatabase().getCollection(collectionName);
+        this.coll = JacksonDBCollection.wrap(dbCollection, AlertImpl.class, String.class, mapperProvider.get());
     }
 
     @Override
-    public Alert factory(AlertCondition.CheckResult checkResult) {
-        if (!checkResult.isTriggered()) {
-            throw new RuntimeException("Tried to create alert from not triggered alert condition result.");
-        }
-
-        Map<String, Object> fields = Maps.newHashMap();
-        fields.put("triggered_at", checkResult.getTriggeredAt());
-        fields.put("condition_id", checkResult.getTriggeredCondition().getId());
-        fields.put("stream_id", checkResult.getTriggeredCondition().getStream().getId());
-        fields.put("description", checkResult.getResultDescription());
-        fields.put("condition_parameters", checkResult.getTriggeredCondition().getParameters());
+    public Alert.Builder builder() {
+        return AlertImpl.builder();
+    }
 
-        return new AlertImpl(fields);
+    @Override
+    public Alert factory(AlertCondition.CheckResult checkResult) {
+        return builder()
+            .streamId(checkResult.getTriggeredCondition().getStream().getId())
+            .conditionId(checkResult.getTriggeredCondition().getId())
+            .description(checkResult.getResultDescription())
+            .conditionParameters(ImmutableMap.copyOf(checkResult.getTriggeredCondition().getParameters()))
+            .triggeredAt(checkResult.getTriggeredAt())
+            .build();
     }
 
     @Override
     public List<Alert> loadRecentOfStream(String streamId, DateTime since) {
-        QueryBuilder qb = QueryBuilder.start("stream_id").is(streamId);
-
-        if (since != null) {
-            qb.and("triggered_at").greaterThanEquals(since.toDate());
-        }
-
-        BasicDBObject sort = new BasicDBObject("triggered_at", -1);
-
-        final List<DBObject> alertObjects = query(AlertImpl.class,
-            qb.get(),
-            sort,
-            AlertImpl.MAX_LIST_COUNT,
-            0
-        );
-
-        List<Alert> alerts = Lists.newArrayList();
-
-        for (DBObject alertObj : alertObjects) {
-            alerts.add(new AlertImpl(new ObjectId(alertObj.get("_id").toString()), alertObj.toMap()));
-        }
-
-        return alerts;
+        return this.coll.find(
+            DBQuery.and(
+                DBQuery.is(AlertImpl.FIELD_STREAM_ID, streamId),
+                DBQuery.greaterThanEquals(AlertImpl.FIELD_TRIGGERED_AT, since.toDate())
+            )
+        )
+            .limit(MAX_LIST_COUNT)
+            .sort(DBSort.desc(AlertImpl.FIELD_TRIGGERED_AT))
+            .toArray()
+            .stream()
+            .collect(Collectors.toList());
```

Comment:
As much as I like the new Stream API in Java 8, it is unnecessary here.

```java
return Collections.unmodifiableList(this.coll.find(
            DBQuery.and(
                DBQuery.is(AlertImpl.FIELD_STREAM_ID, streamId),
                DBQuery.greaterThanEquals(AlertImpl.FIELD_TRIGGERED_AT, since.toDate())
            )
        )
            .limit(MAX_LIST_COUNT)
            .sort(DBSort.desc(AlertImpl.FIELD_TRIGGERED_AT))
            .toArray())
```

If `Alert` and `AlertImpl` were merged, the call to `Collections.unmodifiableList()` to "cast" the result would also be unnecessary.